Qiao En iP

qiaoen17 opened the pull request:
# Kiwi 🥝 

> The First Time I Saw A Kiwi I Thought It's A Potato With Fur - quotedb.org 🥝 

## Kiwi frees your mind of having to remember things you need to do. It's

- text based
- easy to learn
- ~~FAST!!~~ _SUPER FAST!!_ to use

## All you need to do is,

1. download it from [here](https://github.com/qiaoen17/ip).
2. double-click it.
3. add your tasks
4. let it manage it for you 🤠 

## And it is **FREE**

Features:

- [x] Managing Tasks
- [x] Managing deadlines (is it here already?)
- [ ] Reminders (coming soon)

## If you are a Java Programmer, you can use it to practice Java too. Here's the `main` method:
```
public class Main {
    public static void main(String[ ] args) {
            Application.launch(MainApp.class, args);
    }
}
```

Commit:
Add Gradle support

qiaoen17 committed:
Level-1

qiaoen17 committed:
Added List Feature

qiaoen17 committed:
Added Task Class and unmark, mark feature. Possible bug: too specific for detecting mark and unmark.

qiaoen17 committed:
Added Deadline, Event, ToDo classes

qiaoen17 committed:
Added DukeException, handled errors, more errors to be handled later

qiaoen17 committed:
Added handling of index out of bounds for marking and unmarking tasks

qiaoen17 committed:
Added delete feature, updated list feature to update indexing

qiaoen17 committed:
Added enums, will implement if applicable in the future

qiaoen17 committed:
Added java doc comments

qiaoen17 committed:
Add Save function to Duke.java

qiaoen17 committed:
Add comments for toString() in Task.java, Todo.java, Deadline.java and Event.java

qiaoen17 committed:
Add something trivial

qiaoen17 committed:
Merge branch 'master' into branch-Level-7

qiaoen17 committed:
Delete something trivial

qiaoen17 committed:
Revert "Delete something trivial"

This reverts commit c6af468422588d5ad32960a03006c1ecc7521ebf.

qiaoen17 committed:
Delete something trivial

qiaoen17 committed:
Merge branch 'master' into branch-Level-7

qiaoen17 committed:
Add Parser.java, add LocalDate and DateFormatter

qiaoen17 committed:
Add random for branching

qiaoen17 committed:
Merge branch 'master' into branch-Level-8

qiaoen17 committed:
Merge branch 'master' into branch-Level-7

qiaoen17 committed:
Resolved conflicts

qiaoen17 committed:
Resolved conflicts

qiaoen17 committed:
Added Parser.java, Storage.java, TaskList.java, Ui.java

qiaoen17 committed:
Move files into Package duke

qiaoen17 committed:
Merge branch 'master' into add-gradle-support

qiaoen17 committed:
Add final keyword to date and time in Deadline.java

qiaoen17 committed:
Merge branch 'master' into add-gradle-support

qiaoen17 committed:
Add ParserTest.java and ToDoTest.java

qiaoen17 committed:
Create a JAR File

qiaoen17 committed:
Added JavaDoc comments

qiaoen17 committed:
Change to meet coding standards

qiaoen17 committed:
Add find feature to Parser.java

qiaoen17 committed:
Fixed logo in Ui.java

qiaoen17 committed:
Merge branch 'master' into branch-A-JavaDoc

qiaoen17 committed:
Merge branch 'master' into branch-A-CodingStandard

qiaoen17 committed:
Merge branch 'master' into branch-Level-9

# Conflicts:
#	src/main/java/duke/Ui.java

qiaoen17 committed:
Merge branch 'branch-A-CodingStandard'

# Conflicts:
#	src/main/java/duke/Deadline.java

qiaoen17 committed:
Merge branch 'master' into branch-Level-9

rama-pang commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,117 @@
+     */
+    @Override
+    public void printAdded() {
+        System.out.println(Ui.straightLine + "\n  Yep, it's in!\n    [D][ ] " + this.getDescription() + " (by: " +
+                            date.format(DateTimeFormatter.ofPattern("d MMM yyyy")) + " " + time + ")\n  " +
+                            this.getIndex() + " tasks left, 頑張れ!\n" + Ui.straightLine + "\n");
```
Perhaps the line break should be before the operator (e.g. the `+` operator), according to [Java coding standards](https://se-education.org/guides/conventions/java/intermediate.html)
```suggestion
        System.out.println(Ui.straightLine + "\n  Yep, it's in!\n    [D][ ] " + this.getDescription() + " (by: "
                            + date.format(DateTimeFormatter.ofPattern("d MMM yyyy")) + " " + time + ")\n  "
                            + this.getIndex() + " tasks left, 頑張れ!\n" + Ui.straightLine + "\n");
```
I noticed the same issue in several other places too.

rama-pang commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,135 @@
+     */
+    public void printInsufficientInfoException(Keywords keyword) {
+        switch (keyword) {
+            case delete:
```
According to [Java coding standards](https://se-education.org/guides/conventions/java/intermediate.html), there should be no indentation for `case` clauses.

qiaoen17 replied:
Yep, fixed that already! Thank you for the tip!

rama-pang commented on src/main/java/duke/ToDo.java:
```diff
@@ -0,0 +1,97 @@
+    }
+
+    @Override
+    public String savedString() {
```
Perhaps the method name could be renamed using a verb, according to [this](https://nus-cs2103-ay2223s1.github.io/website/se-book-adapted/chapters/codeQuality.html).

qiaoen17 replied:
Ah okay, I'll be taking that suggestion thanks :)

rama-pang commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,130 @@
+            String input = sc.nextLine();
+
+                if (input.startsWith("T")) {
+                    String desc = descStr(input);
+                    if (String.valueOf(input.charAt(1)) == "Y") {
+                        tasks.add(new ToDo(true, desc, index));
+                    } else {
+                        tasks.add(new ToDo(false, desc, index));
+                    }
+                    index++;
+                    continue;
+                }
+                if (input.startsWith("D")) {
+                    String desc = descStr(input);
+                    String time = timeStr(input);
+                    LocalDate date = dateStr(input);
+                    if (String.valueOf(input.charAt(1)) == "Y") {
+                        tasks.add(new Deadline(true, desc, index, date, time));
+                    } else {
+                        tasks.add(new Deadline(false, desc, index, date, time));
+                    }
+                    index++;
+                    continue;
+                }
+                if (input.startsWith("E")) {
+                    String desc = descStr(input);
+                    String time = timeStr(input);
+                    LocalDate date = dateStr(input);
+                    if (String.valueOf(input.charAt(1)) == "Y") {
+                        tasks.add(new Event(true, desc, index, date, time));
+                    } else {
+                        tasks.add(new Event(false, desc, index, date, time));
+                    }
+                        index++;
+                        continue;
+                    }
+                }
+
+                sc.close();
+                return new TaskList(tasks, new Ui());
```
Perhaps the indentation should be 4 spaces instead? I noticed the same indentation issue in some other places too.

qiaoen17 replied:
Noted, thank you very much!

rama-pang commented on src/main/java/duke/TaskList.java:
```diff
@@ -0,0 +1,111 @@
+        t.markUndone();
+    }
+
+    public void findMatchingTask(String keyword) {
```
Perhaps the method name should be `findMatchingTasks` instead, as the matching tasks is plural?

rama-pang commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,135 @@
+     */
+    public Ui() {}
+
+    public static String straightLine = "  ----------------------------------------------------------------------------------";
```
I think this line exceeds the 120 character per line hard limit.

qiaoen17 replied:
Fixed the issue by removing in a later update :,)

rama-pang commented on src/main/java/duke/Deadline.java:
```diff
@@ -0,0 +1,117 @@
+    }
+
+    /**
+     * Output the deadline added to the list of tasks in the console.
```
I think there's a slight grammar mistake here. I noticed the same issue in several other places too.
```suggestion
     * Outputs the deadline added to the list of tasks in the console.
```

qiaoen17 replied:
Changed those to imperative form 👍

rama-pang commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,130 @@
+
+    public void save(TaskList tasks) {
+            File f = new File(filePath);
+            if (!f.exists()) {
+                try {
+                    f.createNewFile();
+                    FileWriter fw = new FileWriter(filePath);
+                    for (int i = 0; i < tasks.getLength(); i++) {
+                        fw.write(tasks.getTask(i).savedString() + System.lineSeparator());
+                    }
+                    fw.close();
+                } catch (IOException e) {
+                    Ui.printCannotOpenFile();
+                }
+            } else {
+                try {
+                    FileWriter fw = new FileWriter(filePath);
+                    for (int i = 0; i < tasks.getLength(); i++) {
+                        fw.write(tasks.getTask(i).savedString() + System.lineSeparator());
+                    }
+                    fw.close();
+                } catch (IOException e) {
+                    Ui.printUnableToAppend();
+                }
+            }
+    }
```
Perhaps the duplicate code could be separated into a method instead? I noticed some duplicate code in other places too.

qiaoen17 replied:
Oo, okay, I'll try that 👍

notle1706 commented on src/main/java/duke/Parser.java:
```diff
@@ -0,0 +1,144 @@
+     * @param time the String input time from console.
+     * @return A String representation of time to be output in console.
+     */
+    public String parseTime(String time) {
```
Perhaps you could make use of the DateTimeFormatter class in java to help with this implementation?

qiaoen17 replied:
Sounds good, let me check how I can weave that in!

notle1706 commented on src/main/java/duke/Storage.java:
```diff
@@ -0,0 +1,130 @@
+                } catch (IOException e) {
+                    Ui.printUnableToAppend();
+                }
+            }
```
Indentation should be level 8 instead of level 12.

notle1706 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,135 @@
+     */
+    public Ui() {}
+
+    public static String straightLine = "  ----------------------------------------------------------------------------------";
```
Perhaps you could change `straightLine` to a static constant, `STRAIGHT_LINE` as your ToDo.java also declared a `straightLine` variable

notle1706 commented on src/main/java/duke/Ui.java:
```diff
@@ -0,0 +1,135 @@
+
+    public static String straightLine = "  ----------------------------------------------------------------------------------";
+
+    public String logo  =         "___   ___    ___  ____                     ___   ___ \n"
+                               + "|  |  /  /  |  |  \\  \\      ------      /  /   |  | \n"
+                               + "|   /   /   |  |   \\  \\    / / \\ \\    /  /    |  | \n"
+                               + "|   \\  \\   |  |     \\  \\  /  /   \\ \\  /  /     |  | \n"
+                               + "|    \\  \\  |  |      \\  \\/  /     \\ \\/  /      |  | \n"
+                               + "|___| \\__\\ |__|       \\____/       \\____/       |__| \n";
```
I think this is very cool and nice. Good effort! 😺

notle1706 reviewed (commented):
LGTM aside from the few comments. Great job!

qiaoen17 committed:
Added checkstyle, tweaked to adhere to more coding standards

limweijun commented on src/main/java/duke/Event.java:
```diff
@@ -0,0 +1,125 @@
+    public void printAdded() {
+        System.out.println(Ui.straightLine + "\n  Looks like you have a new event:\n    [E][ ] " + this.getDescription()
+                            + " (at: " + date.format(DateTimeFormatter.ofPattern("d MMM yyyy")) + " " + time + ")" +
+                            "\n  " + this.getIndex() + " tasks left, 頑張れ!\n" + Ui.straightLine + "\n");
```
As stated by other reviewer, your "+" operator should be at the start of line 61, the same goes to line 71, 74, 87 and other lines of code as well

qiaoen17 replied:
👍 yepp fixed that issue

AkkFiros commented on src/main/java/duke/Duke.java:
```diff
@@ -0,0 +1,48 @@
+
+    }
+
+    public static void run(TaskList taskList, Ui ui, Storage storage) {
```
While there are JavaDocs for most of the methods, it might be helpful to add JavaDocs descriptor for more methods/classes!

qiaoen17 replied:
Noted on that, thank you 👍

AkkFiros reviewed (commented):
Code is generally readable and very well structured! Will help newer users if there were more Java Docs for other classes and methods!

qiaoen17 committed:
Added GUI to duke

qiaoen17 committed:
Add assertions in the code.

No assert statements in the code.

It needs to change so that important assumptions that should hold at
various points in the code is documented.

Add assert statements to Task.java for getDescription and getIndex
methods, and to TaskList.java for the addEvent, addToDo, addDeadline
methods.

Using assertions in these areas ensure the assumption holds for each
method, that description for tasks are never empty, and there is a
task being added into the TaskList every time the add function is
invoked.

Assertions are not only simple, but their impact on performance is
considered low and worth the additional safety they provide.

qiaoen17 committed:
Improve Code Quality

Code Quality is not considered when writing code.

Add new methods to make code more readable through abstraction, rename
methods to fit with Code Quality standards.

To make code more reader friendly.

qiaoen17 committed:
Remove dead code

qiaoen17 committed:
Merge pull request #2 from qiaoen17/branch-A-Assertions

Branch-A-Assertions

qiaoen17 committed:
Merge branch 'master' into branch-A-CodeQuality

qiaoen17 committed:
Merge pull request #3 from qiaoen17/branch-A-CodeQuality

Improve Code Quality

qiaoen17 committed:
Merge branch 'branch-A-CodeQuality'

# Conflicts:
#	src/main/java/duke/TaskList.java

qiaoen17 committed:
Remove spacing in TaskList.java

qiaoen17 committed:
Merge branch 'master' of https://github.com/qiaoen17/ip

qiaoen17 committed:
Add CI for future reference

qiaoen17 committed:
Add C-Tagging function, might expand to saving tags in the file later,
or finding task by tag.

qiaoen17 committed:
Delete saved file

qiaoen17 committed:
Improve GUI

Added background image and shaped text box, as well as prompt text

qiaoen17 committed:
More improvements to GUI, add tags shown beside task

qiaoen17 committed:
Fixed checkstyle root error

qiaoen17 committed:
Fix Jar file error and add Ui picture

qiaoen17 committed:
Add UserGuide

qiaoen17 committed:
Trial fix of unable to generate text file on other computers

qiaoen17 committed:
Clean dead code and add comments
